Menus/InventoryMenuGUI.cpp

```diff
@@ -0,0 +1,1237 @@
+#include "InventoryMenuGUI.h"
+
+#include "FrameMan.h"
+#include "UInputMan.h"
+#include "MovableMan.h"
+#include "PresetMan.h"
+
+#include "Controller.h"
+#include "AHuman.h"
+#include "HDFirearm.h"
+#include "Icon.h"
+
+#include "AllegroBitmap.h"
+#include "GUIFont.h"
+
+#include "GUIControlManager.h"
+#include "GUICollectionBox.h"
+#include "GUILabel.h"
+#include "GUIButton.h"
+#include "GUIScrollbar.h"
+
+namespace RTE {
+
+	const Vector InventoryMenuGUI::c_CarouselBoxMaxSize(50, 32);
+	const Vector InventoryMenuGUI::c_CarouselBoxMinSize(40, 24);
+	const Vector InventoryMenuGUI::c_CarouselBoxSizeStep = (c_CarouselBoxMaxSize - c_CarouselBoxMinSize) / (c_ItemsPerRow / 2);
+	const int InventoryMenuGUI::c_CarouselBoxCornerRadius = ((c_CarouselBoxMaxSize.GetFloorIntY() - c_CarouselBoxMinSize.GetFloorIntY()) / 2) - 1;
+
+	BITMAP *InventoryMenuGUI::s_CursorBitmap = nullptr;
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	void InventoryMenuGUI::CarouselItemBox::GetIconsAndMass(std::vector<BITMAP *> &itemIcons, float &totalItemMass, const std::vector<MovableObject *> *equippedItems) const {
+		if (IsForEquippedItems) {
+			itemIcons.reserve(equippedItems->size());
+			for (const MovableObject *equippedItem : *equippedItems) {
+				if (equippedItem) {
+					itemIcons.push_back(equippedItem->GetGraphicalIcon());
+					totalItemMass += equippedItem->GetMass();
+				}
+			}
+		} else {
+			itemIcons.push_back(Item->GetGraphicalIcon());
+			totalItemMass += Item->GetMass();
+		}
+	}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	void InventoryMenuGUI::Clear() {
+		m_SmallFont = nullptr;
+		m_LargeFont = nullptr;
+
+		m_MenuController = nullptr;
+		m_InventoryActor = nullptr;
+		m_MenuMode = MenuMode::Carousel;
+		m_CenterPos.Reset();
+
+		m_EnabledState = EnabledState::Disabled;
+		m_EnableDisableAnimationTimer.Reset();
+		m_EnableDisableAnimationTimer.SetRealTimeLimitMS(300);
+
+		m_InventoryActorIsHuman = false;
+		m_InventoryActorEquippedItems.clear();
+
+		m_CarouselDrawEmptyBoxes = true;
+		m_CarouselBackgroundTransparent = true;
+		m_CarouselBackgroundBoxColor = 4;
+		m_CarouselBackgroundBoxBorderSize = Vector(2, 1);
+		m_CarouselBackgroundBoxBorderColor = g_MaskColor;
+
+		m_CarouselAnimationDirection = CarouselAnimationDirection::None;
+		m_CarouselAnimationTimer.Reset();
+		m_CarouselAnimationTimer.SetRealTimeLimitMS(200);
+		for (std::unique_ptr<CarouselItemBox> &carouselItemBox : m_CarouselItemBoxes) { carouselItemBox = nullptr; }
```

Comment:
Fill `nullptr` will work fine with unique_ptrs right? I.e. it'll count as setting them to null and delete appropriately?
As for why I'm using nullptrs, I guess I didn't wanna deal with adding some reset capability to the struct, since that'd probably be needed to use object directly. Tbh though, I think really I just did it cause I wanted to lol

Reply:
Fill is pretty much exactly what you're doing here.
And tbh i'd rather you just added a Reset.
Using uniques here makes it seem more complicated than it is.

Reply:
Yeah okay, fill doesn't work with unique_ptrs. I may have tried that before and had to not use it cause of this:
![image](https://user-images.githubusercontent.com/7739094/121837220-6facc680-ccab-11eb-8725-c16acdb73a9a.png)

Also, how does unique_ptr here make it more complicated? Unique_ptr is much the same as object, except it doesn't need a reset since you can just delet it. I don't particularly mind changing it, I just don't really get why you want it changed.

Reply:
I don't particularly want it changed, it just feels like a hack.
If I see a smort_ptr I just assume there's some dynamic allocation that needs managing going on, rather then a cleverass pseudoreset.
Also muh gainzz wasted on destructing and constructing.

Reply:
Okay, I'll leave it be then. The gainz problem is minor since these things aren't actually destructed during regular use, just when the InventoryMenuGUI is deded, so it's not really affecting much.
I dunno if it's really a hack, I could use non smort pointers to do the same thing, just I'd have to actively delete them on destroy. I'm not really sure if cpp would rather you use objects or smort_ptrs nowadays, I feel like it's the latter though. Their use case is different sometimes, but overlaps at other times (e.g. this).